### PR TITLE
Direct local on statements

### PR DESCRIPTION
**Pull request by mppf**
Before this change, the compiler always allocated an argument bundle for
an on statement execution even if that on statement was to run locally.

After this change, instead of unconditionally emitting code to allocate
the argument bundle and call chpl_executeOn, the compiler emits a code
sequence like this:

    if (chpl_doDirectExecuteOn(targetLocale)) {
       on_body(args...);
    } else {
       // the old sequence, as before
       allocate arg bundle
       (use autoCopy if necessary to copy into arg bundle)
       // call wrap_on_body through the runtime
       chpl_executeOn(targetLocale, arg bundle, wrap_on_body)
       free arg bundle
    }

The chpl_doDirectExecuteOn function is implemented in the locale models.
It needs to be provided by the locale model because it is up to the
locale model what the sublocale values mean and it is up to the locale
model what the policy is about when a task needs to be moved to another
locale/sublocale.

Additionally, the previous implementation tried to add autoDestroy calls
before downEndCount in the task/on function. That presented a serious
problem with the above pattern, since the matching autoCopy call is only
generated if the arg bundle is allocated and the wrapper is called. While
there are other ways to solve this problem, I chose to move these
autoDestroy calls and the downEndCount call to the wrapper function. That
approach has these advantages:

  * it allows the direct on optimization to work
  * it allows similar optimization for tasks (e.g. if the compiler
    emitted code to check if a task should actually be spawned or if the
    body should just be run. Running the body directly might be useful in
    case there are already too many tasks waiting to be run for example)
  * it separates the concerns of the autoDestroy and the downEndCount
    from the task/on body
  * it ensures that the autoDestroy calls are always run before the
    downEndCount (while the previous implementation sometimes did not
    meet this property).

The downEndCount is initially added very early in compilation (in
build.cpp). While in principle the downEndCount calls could be added much
later and added directly to the wrapper function, I chose to move them
from the task/on function into the wrapper in parallel.cpp. That
minimizes changes to other parts of the compiler.

This approach also enabled me to remove insertBeforeDownEndCount, which
was buggy, and to replace it with more accurate code that is only used in
parallel.cpp.
	
This PR also updates the fences added for --cache-remote to be in the
wrapper function for such on statements in order to avoid these fences
when the on statement amounts to a local function call.

Passed full local testing.
Passed full GASNet+fifo testing.
Passed release/examples numa testing.

Reviewed by @benharsh - thanks!

### CODE REVIEW

**Comment by mppf** [comment]
This PR should be merged after #3504.

**Comment by mppf** [comment]
Full local testing passed except for trivial/bradc/wideref2string.
GASNet testing had 96 failures including trivial/bradc/ref2string, associative/bharshbarg/arrays/opEquals, associative/bharshbarg/domains/reduceArrOfAssocDom.

**Comment by mppf** [comment]
@benharsh - I'm pretty sure this will pass testing now, but I'm re-running to be sure. It's a good time for you to have a look. Thanks!

@ronawho - have a look at how I changed downEndCount in case it is related to what you are doing.

File: modules/internal/localeModels/numa/LocaleModel.chpl [hunk 1]
```diff
@@ -440,6 +440,35 @@ module LocaleModel {
...
+  //  else
+  //         chpl_executeOn / chpl_executeOnFast
+  //
+  export
+  proc chpl_doDirectExecuteOn(loc: chpl_localeID_t // target locale
+                             ):bool {
+    const dnode =  chpl_nodeFromLocaleID(loc);
+    const dsubloc =  chpl_sublocFromLocaleID(loc);
+
+    if (done != chpl_nodeID) {
```
Comment (benharsh) [inline modules/internal/localeModels/numa/LocaleModel.chpl]: Is this supposed to be 'dnode' ?

**Comment by benharsh** [comment]
I don't entirely grok the --cache-remote or EndCount changes, but the rest looks sane to me. I think it would be good to do a little numa testing, though I wouldn't require the entire suite.

Do you have a sense of which benchmarks this might help?

Looks like a good change to me!

File: modules/internal/localeModels/numa/LocaleModel.chpl (hunk 1 above)
Comment (mppf) [inline modules/internal/localeModels/numa/LocaleModel.chpl]: Good catch, and a reminder for me to run some numa tests.

**Comment by mppf** [comment]
@benharsh - no, I don't have a sense of which benchmarks it will help, but I definitely saw improvements for labelprop-tweets. I'm pretty sure that I/O intensive code will benefit a lot. I'll do NUMA testing and then plan to merge it on Monday.

File: modules/internal/localeModels/flat/LocaleModel.chpl [hunk 2]
```diff
@@ -358,6 +358,24 @@ module LocaleModel {
   extern proc chpl_ftable_call(fn: int, args: c_void_ptr): void;
 
   //
+  // returns true if an executeOn can be handled directly
+  // by running the function in question.
+  // Applieds to execute on and execute on fast.
```
Comment (ronawho) [inline modules/internal/localeModels/flat/LocaleModel.chpl]: Applieds? (sp?)

**Comment by ronawho** [comment]
I took just enough of a look to see that this is unrelated to the endcount stuff I'm working on. Let me know if you'd like me to give it a more thorough review.

**Pull request merged**